src/mlpack/methods/ann/layer/upsample_impl.hpp

```diff
@@ -0,0 +1,275 @@
+/**
+ * @file methods/ann/layer/upsample_impl.hpp
+ * @author Abhinav Anand
+ *
+ * Implementation of the Upsample layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_UPSAMPLE_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_UPSAMPLE_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "upsample.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename InputDataType, typename OutputDataType>
+  Upsample<InputDataType, OutputDataType>::
+  Upsample():
+  inRowSize(0),
+  inColSize(0),
+  outRowSize(0),
+  outColSize(0),
+  depth(0),
+  mode("nearest"),
+  batchSize(0)
+  {
+  // Nothing to do here.
+  }
+
+template<typename InputDataType, typename OutputDataType>
+  Upsample<InputDataType, OutputDataType>::
+  Upsample(
+    const size_t inRowSize,
+    const size_t inColSize,
+    const size_t outRowSize,
+    const size_t outColSize,
+    const size_t depth,
+    const string mode):
+  inRowSize(inRowSize),
+  inColSize(inColSize),
+  outRowSize(outRowSize),
+  outColSize(outColSize),
+  depth(depth),
+  mode(mode),
+  batchSize(0)
+  {
+  // Nothing to do here.
+  }
+
+template<typename InputDataType, typename OutputDataType>
+  void Upsample<InputDataType, OutputDataType>::Reset()
+  {
+    if(mode == "bilinear")
```

Comment:
Wondering if it might be useful to implement a separate layer for each interpolation strategy, let me know what you think.

Reply:
Yes I think its better to make separate layer for these three interpolations. In one layer it getting a bit complicated to implement forward and backward pass. Currently there is a layer called `bilinear interpolation` but its implementation seems wrong, it does not match with pytorch outputs.
So I will make three different layers called `nearest_interpolation`, `bilinear_interpolation`(fix) and `bicubic_interpolation`.